File: scripts/run_smk.R
library(trend)

base_dir <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
file_path <- file.path(base_dir, "data", "GEE_data", "computed_results", "national_monthly_water_area_2015_2025.csv")
df <- read.csv(file_path)

# Full monthly dataset
df_full <- df

# Filter July for annual peak trend analysis
df_july <- subset(df, month == 7)

# Sort chronologically
df_july <- df_july[order(df_july$year), ]

cat("Total months:", nrow(df_full), "\n")
cat("July observations:", nrow(df_july), "\n")

# Create time series for full monthly data (for seasonal MK)
df_full <- df_full[order(df_full$year, df_full$month), ]
ts_data <- ts(df_full$water_area_calibrated_km2, start = c(2015, 1), frequency = 12)

# Perform Seasonal Mann-Kendall Test on full monthly series
smk <- smk.test(ts_data)
cat("\n--- Seasonal Mann-Kendall Test (Full Monthly) ---\n")
print(smk)

# Sen's Slope on full monthly series
sen <- sens.slope(ts_data)
cat("\n--- Sen's Slope (Full Monthly) ---\n")
print(sen)
cat("Sen's Slope (Annualized):", sen$estimates * 12, "km2/year\n")

# Mann-Kendall on July peak series (n=11)
cat("\n--- Mann-Kendall Test (July Peak, n=11) ---\n")
july_vals <- df_july$water_area_calibrated_km2
mk_july <- mk.test(july_vals)
print(mk_july)

sen_july <- sens.slope(ts(july_vals, start = 2015, frequency = 1))
cat("\nSen's Slope (July):", sen_july$estimates, "km2/year\n")

File: scripts/run_smk.py
import pandas as pd
import numpy as np
import pymannkendall as mk
import warnings
warnings.filterwarnings('ignore')

file_path = r"D:\Drubo_IWm\Drubo_all\Project\Publication\Project_HydroSAR-Bangladesh\SAR Analysis Paper\data\GEE_data\Final_Interpolated_Master_Dataset_2015_2025.csv"
df = pd.read_csv(file_path)

# Filter National Data
df_nat = df[df['Scope'] == 'National']

# Sort chronologically
df_nat = df_nat.sort_values(by=['Year', 'Month'])

# Display the series
print(f"Total months: {len(df_nat)}")

# Create a time series list
time_series = df_nat['Area_km2'].tolist()

try:
    # Perform Seasonal Mann-Kendall Test (period=12 for monthly data)
    res_smk = mk.seasonal_test(time_series, period=12)
    print("\n--- Seasonal Mann-Kendall Test ---")
    print(f"Trend: {res_smk.trend}")
    print(f"p-value: {res_smk.p}")
    print(f"Z-Score: {res_smk.z}")
    print(f"Tau: {res_smk.Tau}")
    print(f"Sen's Slope: {res_smk.slope} km2/month")
except Exception as e:
    print(f"Pymannkendall error: {e}")

try:
    # Let's also do a simple Sen's slope with confidence intervals using scipy or statsmodels if needed
    import scipy.stats as stats
    x = np.arange(len(time_series))
    slope, intercept, r_value, p_value, std_err = stats.linregress(x, time_series)
    print(f"\nLinear Regression Slope (for reference): {slope*12:.2f} km2/year, p={p_value:.4f}")
except Exception as e:
    pass
